### PR TITLE
Multi-tensor SGDOptimizer (on device training)

### PR DESCRIPTION
**Pull request by pengwa**
Implement SGDOptimizerV2 taking sequence of weights and gradients as inputs.

For CPU EP and CUDA EP only.

Added tests.

### CODE REVIEW

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 1]
```diff
@@ -1118,6 +1118,46 @@ void RegisterTrainingOpSchemas() {
           {"float"},
           "Constrain learning rate to float");
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(SGDOptimizerV2)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "lr", "The learning rate.", "T1")
+      .Input(1, "weights", "Sequence of weights to optimize.", "S_WEIGHT")
+      .Input(2, "gradients", "Sequence of gradients computed in this iteration.", "S_GRAD")
+      .Input(3, "update_signal",
```
Comment (askhade) [inline orttraining/orttraining/core/graph/training_op_defs.cc:1127]: nit: rename to skip_update to make it more intuitive?

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 2]
```diff
@@ -1118,6 +1118,46 @@ void RegisterTrainingOpSchemas() {
...
+             "This signal indicates if weight updates are skipped, applicable to gradient infinity check"
+             " in mixed precision training. ",
+             "T_BOOL", OpSchema::Optional)
+      .Output(0, "updated_flag", "Whether gradient is applied or not.", "T_BOOL")
```
Comment (askhade) [inline orttraining/orttraining/core/graph/training_op_defs.cc:1131]: nit: rename to update_completed to make it more intuitive?

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 3]
```diff
@@ -1118,6 +1118,46 @@ void RegisterTrainingOpSchemas() {
           {"float"},
           "Constrain learning rate to float");
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(SGDOptimizerV2)
+      .SetDomain(kMSDomain)
```
Comment (askhade) [inline orttraining/orttraining/core/graph/training_op_defs.cc:1122]: Please add some description for the op.

File: orttraining/orttraining/training_ops/cpu/optimizer/common.cc [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/providers/cpu/tensor/utils.h"
+#include "orttraining/training_ops/cpu/optimizer/common.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+Status CopyIfNotSameCPUBuffer(OpKernelContext* ctx, size_t number_of_values,
+                              const TensorSeq* values, TensorSeq* updated_values) {
```
Comment (askhade) [inline orttraining/orttraining/training_ops/cpu/optimizer/common.cc:13]: nit: src or src_values and dest and dest_values is a more intuitive name

File: orttraining/orttraining/training_ops/cuda/optimizer/common.h [hunk 5]
```diff
@@ -3,11 +3,21 @@
 
 #pragma once
 #include "core/providers/cuda/cuda_common.h"
-#include "orttraining/training_ops/cpu/optimizer/common.h"
 
 namespace onnxruntime {
 namespace cuda {
 
+template <typename TC, typename TS>
+TC compute_bias_correction_coefficient(
```
Comment (askhade) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.h:11]: why duplicate this? why not use the one in onnxruntime::contrib?

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc:1139]: oh, maybe the comment is confusing, update_signal means not to skip, so we cannot use skip_update. I have update the comment to make it clear.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 3 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc:1134]: Forgot this, done. :)

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc]: done

File: orttraining/orttraining/training_ops/cuda/optimizer/common.h (hunk 5 above)
Comment (pengwa) [reply orttraining/orttraining/training_ops/cuda/optimizer/common.h]: You are right. Removed this duplication in cuda path

**Review by askhade — APPROVED** [review]
LGTM Thanks!

**Pull request merged**